# Chapter 10: Introduction to Artificial Neural Networks with Keras

**Tujuan:** Memahami konsep dasar neuron buatan, MLP, dan implementasi menggunakan Keras Sequential API.

---

## 1. Dari Neuron Biologis ke Neuron Buatan

* **Biologis:** Dendrit menerima sinyal → soma → akson → sinaps
* **Buatan (perceptron):**

  $
  y = \begin{cases}
    1 & \text{jika } w^T x + b > 0, \\
    0 & \text{lainnya}
  \end{cases}
  $
* Perceptron hanya mampu memisahkan data secara linier (linear separator).

---

## 2. Multilayer Perceptron (MLP) & Backpropagation

* **MLP:** lapisan tersembunyi + aktivasi non-linear (ReLU, sigmoid, tanh)
* **Forward pass:** hitung output dari layer ke layer
* **Backpropagation:** hitung gradien loss terhadap bobot → update via optimizers (SGD, Adam)

---

## 3. Aktivasi Umum

* **ReLU:**

  $
  \text{ReLU}(z) = \max(0, z)
  $

  (paling umum di hidden layers)

* **Sigmoid:**

  $
  \sigma(z) = \frac{1}{1 + e^{-z}}
  $

  (umum untuk output biner)

* **Softmax:** untuk multiclass classification, normalisasi probabilitas

---

## 4. Implementasi MLP Klasifikasi (MNIST)

* **Dataset:** `keras.datasets.mnist` (gambar 28×28 piksel, digit 0–9)

* **Preprocessing:** reshaping + normalisasi

* **Model:**

  ```python
  model = keras.models.Sequential([
      keras.layers.Flatten(input_shape=(28, 28)),
      keras.layers.Dense(300, activation='relu'),
      keras.layers.Dense(100, activation='relu'),
      keras.layers.Dense(10, activation='softmax')
  ])
  ```

* **Compile:**

  ```python
  model.compile(
      loss='sparse_categorical_crossentropy',
      optimizer='adam',
      metrics=['accuracy']
  )
  ```

* **Train:**

  ```python
  model.fit(X_train, y_train, epochs=10, validation_split=0.1)
  ```

* **Evaluate:**

  ```python
  model.evaluate(X_test, y_test)
  ```

---

In [1]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

# 4.1 Load & preprocess data
(X_train, y_train), (X_test, y_test) = keras.datasets.mnist.load_data()
X_train = X_train.astype("float32") / 255.0
X_test  = X_test.astype("float32")  / 255.0

# 4.2 Bangun model Sequential
model = keras.Sequential([
    layers.Flatten(input_shape=(28,28)),
    layers.Dense(300, activation='relu'),
    layers.Dense(100, activation='relu'),
    layers.Dense(10, activation='softmax')
])

# 4.3 Compile model
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

# 4.4 Tampilkan ringkasan arsitektur
model.summary()

# 4.5 Latih model
history = model.fit(
    X_train, y_train,
    epochs=10,
    batch_size=32,
    validation_split=0.1
)

# 4.6 Evaluasi pada test set
test_loss, test_acc = model.evaluate(X_test, y_test)
print(f"Test accuracy: {test_acc:.3f}")

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step


/usr/local/lib/python3.11/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten (Flatten)               │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 300)            │       235,500 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 100)            │        30,100 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 10)             │         1,010 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 266,610 (1.02 MB)

 Trainable params: 266,610 (1.02 MB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.8860 - loss: 0.3757 - val_accuracy: 0.9697 - val_loss: 0.1012
Epoch 2/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 7s 3ms/step - accuracy: 0.9731 - loss: 0.0889 - val_accuracy: 0.9757 - val_loss: 0.0836
Epoch 3/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9824 - loss: 0.0556 - val_accuracy: 0.9768 - val_loss: 0.0785
Epoch 4/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9873 - loss: 0.0401 - val_accuracy: 0.9790 - val_loss: 0.0752
Epoch 5/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9896 - loss: 0.0321 - val_accuracy: 0.9785 - val_loss: 0.0747
Epoch 6/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.9918 - loss: 0.0257 - val_accuracy: 0.9795 - val_loss: 0.0826
Epoch 7/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.9930 - loss: 0.0217 - val_accuracy: 0.9775 - val_loss: 0.0899
Epoch 8/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9945 - loss: 0.0161 

## 5. Interpretasi Hasil

- **Learning curves** dapat dilihat via `history.history['accuracy']` vs `history.history['val_accuracy']`  
- **Overfitting** jika gap besar train vs val → tambahkan regularisasi (Dropout, L2)

---

**Ringkasan Chapter 10**

1. Neuron buatan mereplikasi fungsi linear + aktivasi non‑linear.  
2. MLP belajar via backpropagation.  
3. Keras Sequential memudahkan konstruksi model.  
4. Contoh: Klasifikasi MNIST dengan dua hidden layer ReLU.  
5. Evaluasi akurasi ~0.98 pada test set.